packages/edit-post/src/components/secondary-sidebar/inserter-sidebar.js

```diff
@@ -30,6 +31,7 @@ export default function InserterSidebar() {
 	const isMobileViewport = useViewportMatch( 'medium', '<' );
 	const [ inserterDialogRef, inserterDialogProps ] = useDialog( {
 		onClose: () => setIsInserterOpened( false ),
+		focusOnMount: 'secondElement',
```

Comment:
Is this still needed? It feels very specific to have a "secondElement" option in focusOnMount. The reasoning is that the developer might change what's inside a dialog without touching this call at all (two separate concerns), so basically "secondElement" is meaningless. 

In other words, if we really want to target a specific element, we should probably use a dedicate `useEffect(() => { ref.focus() } )` instead of trying to include this logic into a reusable hook.

Reply:
I made this specifically for use in Core because I think it has other good uses. It also had a check to ensure the element existed before focus was attempted. Probably could have done a better job to work on a fallback.

However, I changed to the useEffect approach but this was tricky because I had to figure out a way to only fire useEffect on initial open. It seems useEffect fires every time the component mounts and focus kept happening when I tried to tab out of the search box.

Do you have any suggestions on if this is good enough? Should I try to access edit post store from menu.js directly or is passing a new property fine?

Thanks.

Reply:
> Do you have any suggestions on if this is good enough? Should I try to access edit post store from menu.js directly or is passing a new property fine?

Mmm interesting challenge, I guess neither of these are ideal to me. The Menu shouldn't be aware of an outside `isOpen` because it's not something that is always rendered in a toggable sidebar and block-editor shouldn't have any dependency on edit-post (cycle dependency).

We should try to find an alternative.

Can I ask why we want to focus the "second element" in the first place and not the "first element" like all dialogs?

Reply:
> Can I ask why we want to focus the "second element" in the first place and not the "first element" like all dialogs?

Because in the case of the Block Inserter, it is better to focus the search input first as most users will likely search for their block instead of looking through the list. I know that's how I personally use it. If we were to focus the first element, it would focus the Close button which would create extra key strokes to interact with the search field and the Block Inserter content.

What if I added my changes back to the useFocusOnMount hook with additional checks to make sure the element passed in is valid or default to the first? I understand most devs probably won't use this, but there's a need for it in the Core Gutenberg editor. As you said, passing the state probably isn't a good idea as the Inserter is not always used in a sidebar. Although, I did have one question about that. If the Inserter wasn't used in a sidebar, wouldn't that search field get focus by default since it uses useFocusOnMount hook or would it simply focus the first element on the page? I would think it would focus the first element in the Inserter since the useDialog hook. This would also be super bad UX if the Inserter was just on the page, not in a sidebar, and focus got placed there first. Although, maybe it wouldn't work this way because the Inserter doesn't use those hooks, the sidebar in the editor does.

Personally, I think passing that extra parameter is fine that way we don't need access to core/edit-post store and the variable is set to false, so it's not like it is going to effect other code. Users can still use the useFocusOnMount hook as I've only disabled it in the sidebar managed by Gutenberg. Nothing really changes for the Inserter component itself.

Sorry, this was long winded. Thanks.

Reply:
The focus target for opening a dialog is variable depending on usage. For simple or informational dialogs (e.g., with only one or two focal points), placing focus at the beginning will be best (and ideally that should be the close button). For complex dialogs, however, the usage of the dialog really needs to be taken into consideration. Focusing on the task that needs to be performed using a dialog makes sense. It's still good to place focus as early in the modal as possible, but making it easier to complete the task is a good thing.

Reply:
I agree with @joedolson here, if focus wasn't placed in the search field and I'd first have to tab to it because it landed on a pointless Close button because I just opened the dialog, I'd be pretty annoyed pretty quickly. ;-) So yeah, focus the Search box.

Reply:
Hey folks, thanks for chiming in. Note that I was not arguing about not focusing the input when the dialog opens, I'm just trying to understand how we choose the element to focus when opening dialogs in general. 

From the responses, I sense that it depends on the dialog with a fallback to "first element" as default.

--
Now the problem we have is how to implement this behavior in a way that doesn't mess with our APIs too much (potentially creating maintenance burden).

- The "secondElement" support in the `useFocusOnMount` doesn't sound great because it's just arbitrarily happens that the element we want of focus is second, it can change at any time (say we add another button before the search input)
- Adding `isOpen` to the library component doesn't make sense for me either because that component `Library` (not that Library don't render any dialog) can be used in any place, it's a reusable component and not necessarily a dialog, so `isOpen` prop doesn't make sense there.

What about this potential workaround:

 - Add a class to indify the input:
 - In the parent component rendering the dialog (`InserterSidebar` the one that calls useDialog), have a ref on the wrappe.
 - Add an effect like that to the `InserterSidebar`
 
 ```
useEffect(() => {
    if ( isOpen ) {       sidebarRef.current.querySelector('.some-selector-to-find-search-input' ).focus() 
   }
}, [ isOpen ] }
 ```

WDYT

Reply:
> Note that I was not arguing about not focusing the input when the dialog opens, I'm just trying to understand how we choose the element to focus when opening dialogs in general.

I didn't think you were, I just wanted to make sure this was a necessary change before we went through too much more work. See the request I posted in accessibility channel here.

https://wordpress.slack.com/archives/C02RP4X03/p1639591732063100

I know quite a bit about accessibility, but so do the rest of the team. Just wanted some insight especially since the solution here wasn't so straight forward. I would have hated to do all this work just to find out the Close button is acceptable. Make sense?

Thanks for the suggestion. As I've said a few different times, I'm not a React developer so I probably wouldn't have gotten there myself. I will work on implementing your solution soon as I think it looks great. I guess it makes sense this way to not have to pass a prop all the way to menu.js, now we can simply keep the logic for this in Core. 👍 🙇‍♂️ 🗜️

Reply:
@youknowriad Sadly, your solution did not work. 😞 When I tried to have useEffect in InserterSidebar with a ref on the parent div, it never worked even when passing isOpen as an argument. Whenever I used console.log, useEffect was firing too early for the focus in the search field. I tried to delay useEffect with a timeout before the focus, but never could get that right.

I started reading about children functional components receiving refs and decided to use forwardRef all the way to the search component. This seems highly inefficient though. 😞 

I could use the useEffect inside the search component but it would still require some type of state tracking or else it would never know when the inserter is open. I wish there was a way to unify the elements so it wasn't all separate.

What are your thoughts about the forwardRef approach? It seems bad practice, but it does work and to my knowledge, it could almost be seen as an improvement since now anyone could have control over the search part of the Inserter.

If not this way, can you think of any other ways? I was thinking maybe this could be refactored in to one component, but that kind of defeats the purpose of React. BTW, when I mentioned console.log above, what I meant was in useEffect, console.log fired before my console.log of the Inserter when it was fully loaded. I tried a timeout of 1000 and it wasn't enough.

Reply:
Thanks for giving this a try. I appreciate the efforts.

The forwardRef approach is good practice when the forwarded ref represents the actual component, in the case of the inserter library, I'd expect the forwarded ref to be the inserter container, not just a specific element of that component (the input).

I'll take a look at the PR later today and see if I can find something that would work well.

Reply:
I've pushed a fix to your branch that I think works. Let me know what you think. I also noticed that you had inverted the condition about when to show the close button (should be hidden on desktop).

Last thing, this `InserterSidebar` component is something that is also replicated in three other screens (site editor, widgets screen and navigation screen) so three other packages. I think we might want to make the same changes there as well.

Reply:
@youknowriad Can you check it out now? I also fixed a bug where the Inserter would not close in Widgets view. I adjusted the code in useCallback just a bit, seems to work now.

Reply:
BTW, this is the same approach I took except I didn't add the "input" selector. 😕 One tiny selector away and it would have worked for me. Thanks for the catch on this and ensuring I didn't create unnecessary ref forwarding to get this done.